pylint/plugins/hass_enforce_type_hints.py

```diff
@@ -1282,10 +1279,20 @@ def visit_module(self, node: nodes.Module) -> None:
     def visit_classdef(self, node: nodes.ClassDef) -> None:
         """Called when a ClassDef node is visited."""
         ancestor: nodes.ClassDef
+        matches: list[TypeHintMatch] = []
         for ancestor in node.ancestors():
             for class_matches in self._class_matchers:
                 if ancestor.name == class_matches.base_class:
-                    self._visit_class_functions(node, class_matches.matches)
+                    matches.extend(class_matches.matches)
+
+        # Remove duplicate `device_class` when it is overridden
+        if (
+            _BASE_DEVICE_CLASS in matches
+            and sum(1 for m in matches if m.function_name == "device_class") > 1
+        ):
+            matches.remove(_BASE_DEVICE_CLASS)
```

Comment:
This is specific to just the `device_class` property and wouldn't be easily extendable.
Just a rough idea, but what do you think about something like this?
```py
    def visit_classdef(self, node: nodes.ClassDef) -> None:
        """Called when a ClassDef node is visited."""
        ancestor: nodes.ClassDef
        self._checked_class_method_names: set[str] = set()

        for ancestor in node.ancestors():
            for class_matches in reversed(self._class_matchers):
                if ancestor.name == class_matches.base_class:
                    self._visit_class_functions(node, class_matches.matches)
                    self._checked_class_method_names.update(
                        match.function_name for match in class_matches.matches
                    )

    def _visit_class_functions(
        self, node: nodes.ClassDef, matches: list[TypeHintMatch]
    ) -> None:
        cached_methods: list[nodes.FunctionDef] = list(node.mymethods())

        for match in matches:
            if match.function_name in self._checked_class_method_names:
                continue
            for function_node in cached_methods:
                if match.need_to_check_function(function_node):
                    self._check_function(function_node, match)
```

Note the `reversed` in the second for loop. Furthermore, I've cached `node.mymethods` as it does a lot of isinstance checkes for each match currently.

Reply:
Let me add some tests for `turn_on` and see how it all feels.

Reply:
I've had a play. I couldn't use `match.function_name` because of the special case on `async_step_*` but I was able to move things around and work around the issue using `checked_class_methods.add(function_node.name)`

The `reverse` isn't useful as we don't have duplicates within the same ancestor class, and we are already working the ancestor tree in the correct order.

I have implemented the caching, and improved/cleaned up the tests accordingly.